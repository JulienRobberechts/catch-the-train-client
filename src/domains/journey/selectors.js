export const selectCurrentJourney = (state) => {
  return state?.journey;
};
