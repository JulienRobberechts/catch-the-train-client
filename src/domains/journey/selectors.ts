import { RootState } from "../../redux-store/types";

export const selectCurrentJourney = (state: RootState) => {
  return state?.journey;
};
