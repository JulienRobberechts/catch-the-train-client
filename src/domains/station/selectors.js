export const selectUserConfiguration = (state) =>
  state?.station?.userConfiguration;

export const selectStationConfigurations = (state) =>
  state?.station?.stationConfigurations;
