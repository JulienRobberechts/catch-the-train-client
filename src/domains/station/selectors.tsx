import { RootState } from "../../redux-store/types";

export const selectUserConfiguration = (state: RootState) =>
  state?.station?.userConfiguration;

export const selectStationConfigurations = (state: RootState) =>
  state?.station?.stationConfigurations;
