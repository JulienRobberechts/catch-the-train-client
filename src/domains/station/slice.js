import { createSlice } from "@reduxjs/toolkit";
import { loadStationConfigurations } from "../../adapters/stationPreferences";

export const initialState = {
  userConfiguration: {
    onTimeMarginDelaySeconds: 50,
    timezone: "+01:00",
  },
  stationConfigurations: loadStationConfigurations(),
};

export const slice = createSlice({
  name: "toTheStation",
  initialState,
  reducers: {
    reset: (state) => {
      Object.assign(state, initialState);
    },
    setUserConfiguration: (state, action) => {
      state.userConfiguration = action.payload;
    },
    setStationConfiguration: (state, action) => {
      const { station, ...props } = action.payload || {};
      if (station) {
        state.stationConfigurations[station] = { ...props };
      }
    },
  },
});

export const {
  reset,
  setStationConfiguration,
  setUserConfiguration,
} = slice.actions;

export default slice.reducer;
