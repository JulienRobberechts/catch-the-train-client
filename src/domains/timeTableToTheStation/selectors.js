import moment from "moment";
import { getDelay } from "../toTheStation/pure";

// those selectors are a mashup of timeTable and toTheStation
// in order to provide a ready to use selectors for the timeLine component.

// Be careful with change of references!!!
export const selectEnhancedTimeTable = state => {
  if (!state.timeTable.route || state.toTheStation.noData) return null;

  const { timeTable } = state;
  const nowTime = moment.parseZone(state.toTheStation.currentTime);

  const {
    configuration: { waitingDelaySeconds },
    station: { travelDurationSeconds, onTimeMarginDelaySeconds }
  } = state.toTheStation;

  const trains = state.timeTable.route.trains.map((departure, index) => {
    const departureTime = moment.parseZone(departure.departureTime);
    const departureDuration = moment.duration(departureTime.diff(nowTime));
    const targetTime = moment.parseZone(departure.departureTime);
    const travelDuration = moment.duration({
      seconds: travelDurationSeconds
    });
    const waitingDuration = moment.duration({
      seconds: waitingDelaySeconds
    });

    const { targetDuration, delayDuration, delayStatus } = getDelay({
      nowTime,
      targetTime,
      travelDuration,
      waitingDuration,
      onTimeMarginDelaySeconds
    });

    return {
      index,
      departureTime,
      departureDuration,
      targetDuration,
      delayDuration,
      delayStatus
    };
  });

  return { route: timeTable.route, trains };
};
