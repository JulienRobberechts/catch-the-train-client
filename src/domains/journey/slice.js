import { createSlice } from "@reduxjs/toolkit";

import { getMissions } from "./service";
import { loadJourney } from "../../adapters/journey";

export const initialState = {
  ...loadJourney(),
};

export const slice = createSlice({
  name: "journey",
  initialState,
  reducers: {
    reset: (state) => {
      Object.assign(state, initialState);
    },
    setCurrentJourney: (state, action) => {
      const { network, line, departure, destination } = action.payload;
      const missions = getMissions(departure, destination).join(",");
      state.network = network;
      state.line = line;
      state.departure = departure;
      state.destination = destination;
      state.missions = missions;
    },
  },
});

export const { reset, setCurrentJourney } = slice.actions;

export default slice.reducer;
