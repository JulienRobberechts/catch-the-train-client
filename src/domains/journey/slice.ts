import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { loadJourney } from "../../adapters/journey";
import { FullJourney } from "./types";

export const initialState: FullJourney = {
  ...loadJourney(),
};

export const slice = createSlice({
  name: "journey",
  initialState,
  reducers: {
    reset: (state) => {
      Object.assign(state, initialState);
    },
    setCurrentJourney: (state, action: PayloadAction<FullJourney>) => {
      const {
        network,
        line,
        departure,
        destination,
        missions,
      } = action.payload;
      state.network = network;
      state.line = line;
      state.departure = departure;
      state.destination = destination;
      state.missions = missions;
    },
  },
});

export const { reset, setCurrentJourney } = slice.actions;

export default slice.reducer;
