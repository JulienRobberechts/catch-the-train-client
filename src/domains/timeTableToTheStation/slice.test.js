import { initialState as timeTableInitialState } from "../timeTable/slice";
import { initialState as toTheStationInitialState } from "../toTheStation/slice";
import { selectEnhancedTimeTable } from "./selectors";

const sampleDepartures = [
  {
    trainCode: "0924",
    departureTime: "2020-03-10T09:24:00+01:00",
    mission: "UPAC",
    displayAttributes: "09:24",
    displayDestination: "Cergy-Le-Haut",
  },
  {
    trainCode: "0929",
    departureTime: "2020-03-10T09:29:00+01:00",
    mission: "UPAC",
    displayAttributes: "09:29",
    displayDestination: "Cergy-Le-Haut",
  },
];

describe("slice timeTable", () => {
  const initialRootState = {
    timeTable: timeTableInitialState,
    toTheStation: toTheStationInitialState,
  };
  describe("initial state", () => {
    it("'selectEnhancedTimeTable' should return falsy", () => {
      expect(selectEnhancedTimeTable(initialRootState)).toBeFalsy();
    });
  });
  describe("after initialization", () => {
    const sampleUserConfiguration = {
      onTimeMarginDelaySeconds: 22,
      timezone: "+04:00",
    };
    const sampleStationConfiguration = {
      station: "chatelet+les+halles",
      travelDurationSeconds: 555,
      waitingDelaySeconds: 77,
    };
    const sampleContextRerAChatelet = {
      at: "2020-03-10T09:22:30+01:00",
      provider: "ratp",
      network: "rers",
      line: "a",
      station: {
        name: "Chatelet-Les-Halles",
        slug: "chatelet+les+halles",
      },
      missions: ["ZEBU"],
    };

    const rootState = {
      toTheStation: {
        currentTime: "2020-03-10T09:13:30+01:00",
        currentTrainCode: "0924",
        userConfiguration: sampleUserConfiguration,
        stationConfiguration: sampleStationConfiguration,
      },
      timeTable: {
        loading: false,
        error: null,
        data: {
          context: sampleContextRerAChatelet,
          departures: sampleDepartures,
        },
      },
    };
    it("'selectEnhancedTimeTable' should MatchSnapshot", () => {
      const actualEnhancedTimeTable = selectEnhancedTimeTable(rootState);
      expect(actualEnhancedTimeTable).toBeTruthy();
      expect(actualEnhancedTimeTable.enhancedDepartures.length).toBe(2);
      expect(actualEnhancedTimeTable).toMatchSnapshot();
    });
  });
});
