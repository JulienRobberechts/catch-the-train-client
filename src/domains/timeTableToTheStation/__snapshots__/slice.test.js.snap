// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slice timeTable after initialization 'selectEnhancedTimeTable' should MatchSnapshot 1`] = `
Object {
  "currentDeparture": Object {
    "code": "0924",
    "index": 0,
  },
  "enhancedDepartures": Array [
    Object {
      "delayDuration": "-PT2S",
      "delayStatus": "on-time",
      "departureDuration": "PT10M30S",
      "departureTime": "2020-03-10T08:24:00.000Z",
      "index": 0,
      "trainCode": "0924",
    },
    Object {
      "delayDuration": "PT4M58S",
      "delayStatus": "early",
      "departureDuration": "PT15M30S",
      "departureTime": "2020-03-10T08:29:00.000Z",
      "index": 1,
      "trainCode": "0929",
    },
  ],
  "travel": Object {
    "accessDuration": "PT1M17S",
    "nowTime": "2020-03-10T08:13:30.000Z",
    "travelDuration": "PT9M15S",
  },
}
`;
